File: 494. Target Sum-4.py
# 恰好 target
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        # p
        # s - p
        # p - (s - p) = t
        # 2p = s + t
        # p = (s + t) / 2
        target += sum(nums)
        if target < 0 or target % 2:
            return 0
        target //= 2

        f = [0] * (target + 1)
        f[0] = 1

        for x in nums:
            for c in range(target, x - 1, -1):
                f[c] += f[c - x]

        return f[target]


# 至多 target
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        # p
        # s - p
        # p - (s - p) = t
        # 2p = s + t
        # p = (s + t) / 2
        target += sum(nums)
        if target < 0 or target % 2:
            return 0
        target //= 2

        f = [1] * (target + 1)

        for x in nums:
            for c in range(target, x - 1, -1):
                f[c] += f[c - x]

        return f[target]


# 至少 target
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        # p
        # s - p
        # p - (s - p) = t
        # 2p = s + t
        # p = (s + t) / 2
        target += sum(nums)
        if target < 0 or target % 2:
            return 0
        target //= 2

        f = [0] * (target + 1)
        f[0] = 1

        for x in nums:
            for c in range(target, -1, -1):
                f[c] += f[max(c - x, 0)]

        return f[target]

File: 494. Target Sum-4.cpp
class Solution {
public:
  vector<vector<int>> memo;
  int findTargetSumWays(vector<int> &nums, int target) {
    int n = nums.size();
    int sum = accumulate(nums.begin(), nums.end(), 0);
    if (sum < abs(target) || (sum + target) % 2 != 0)
      return 0;
    int P = (sum + target) / 2;
    memo.assign(n, vector<int>(P + 1, -1));
    return dfs(nums, n - 1, P);
  }

  int dfs(vector<int> &nums, int i, int a) {
    if (i < 0 || a < 0)
      return a == 0 ? 1 : 0;
    int &res = memo[i][a];
    if (res != -1)
      return res;
    return res = dfs(nums, i - 1, a) + dfs(nums, i - 1, a - nums[i]);
  }
};